src/rmp/src/Restructure.cpp

```diff
@@ -0,0 +1,555 @@
+#include "rmp/blif.h"
+#include "rsz/Resizer.hh"
+#include "sta/Graph.hh"
+#include "sta/Liberty.hh"
+#include "sta/Network.hh"
+#include "sta/PathEnd.hh"
+#include "sta/PathExpanded.hh"
+#include "sta/PathRef.hh"
+#include "sta/PatternMatch.hh"
+#include "sta/PortDirection.hh"
+#include "sta/Sdc.hh"
+#include "sta/Search.hh"
+#include "sta/Sta.hh"
+#include "utl/Logger.h"
+
+using utl::RMP;
+
+namespace rmp {
+
+void Restructure::init(ord::OpenRoad* openroad)
+{
+  openroad_ = openroad;
+  logger_ = openroad->getLogger();
+  makeComponents();
+}
+
+void Restructure::makeComponents()
+{
+  db_ = openroad_->getDb();
+}
+
+void Restructure::deleteComponents()
+{
+}
+
+Restructure::~Restructure()
+{
+  deleteComponents();
+}
+
+void Restructure::reset()
+{
+  lib_file_names_.clear();
+  path_insts_.clear();
+}
+
+void Restructure::run(const char* liberty_file_name,
+                      float slack_threshold,
+                      unsigned max_depth)
+{
+  reset();
+  block_ = openroad_->getDb()->getChip()->getBlock();
+  if (!block_)
+    return;
+
+  lib_file_names_.emplace_back(std::string(liberty_file_name));
+  sta::Slack worst_slack = slack_threshold;
+  open_sta_ = openroad_->getSta();
+
+  removeConstCells();
+
+  getBlob(max_depth);
+
+  runABC();
+
+  postABC(worst_slack);
+}
+
+void Restructure::getBlob(unsigned max_depth)
+{
+  rsz::Resizer* resizer = openroad_->getResizer();
+  open_sta_->ensureGraph();
+  open_sta_->ensureLevelized();
+  open_sta_->searchPreamble();
+
+  sta::PinSet ends;
+
+  getEndPoints(ends, opt_mode_ <= Mode::AREA_3, max_depth);
+  if (ends.size()) {
+    sta::PinSet fanin_fanouts = resizer->findFaninFanouts(&ends);
+    // fanin_fanouts.insert(ends.begin(), ends.end()); // Add seq cells
+    logger_->report("Got {} pins in cloud extraction", fanin_fanouts.size());
+    for (sta::Pin* pin : fanin_fanouts) {
+      odb::dbITerm* term = nullptr;
+      odb::dbBTerm* port = nullptr;
+      open_sta_->getDbNetwork()->staToDb(pin, term, port);
+      if (term && term->getInst()->getITerms().size() < 10)
+        path_insts_.insert(term->getInst());
+    }
+    logger_->report("Found {} Insts for restructuring ...", path_insts_.size());
+  }
+}
+
+void Restructure::runABC()
+{
+  input_blif_file_name_
+      = std::string(block_->getConstName()) + "_crit_path.blif";
+  std::vector<std::string> files_to_remove;
+
+  debugPrint(logger_,
+             utl::RMP,
+             "remap",
+             1,
+             "Constants before remap {}",
+             countConsts(block_));
+
+  Blif blif_(openroad_, locell_, loport_, hicell_, hiport_);
+  blif_.setReplaceableInstances(path_insts_);
+  blif_.writeBlif(input_blif_file_name_.c_str());
+  debugPrint(
+      logger_, RMP, "remap", 1, "Writing blif file {}", input_blif_file_name_);
+  files_to_remove.emplace_back(input_blif_file_name_);
+
+  // abc optimization
+  bool area_mode = opt_mode_ <= Mode::AREA_3;
+  std::string abc_script_file = "ord_abc_script.tcl";
```

Comment:
I don't think we should just write into cwd but should allow the user to give a run directory (eg in flows we would use a subdir for the design).